src/ROOTDataSource.cc

```diff
@@ -0,0 +1,401 @@
+    podio::ROOTReader podioReader;
+    podioReader.openFiles(m_filePathList);
+    nEventsInFiles = podioReader.getEntries("events");
+    frame = podio::Frame(podioReader.readEntry("events", 0));
+
+    // Determine over how many events to run
+    if (nEventsInFiles > 0) {
+      /*
+      std::cout << "podio::ROOTDataSource: Found " << nEventsInFiles
+                << " events in files: \n";
+      for (const auto& filePath : m_filePathList) {
+        std::cout << "               - " << filePath << "\n";
+      }
+      */
+    } else {
+      throw std::runtime_error("podio::ROOTDataSource: No events found!");
+    }
+
+    if (nEvents < 0) {
+      m_nEvents = nEventsInFiles;
+    } else if (nEvents == 0) {
+      throw std::runtime_error(
+          "podio::ROOTDataSource: Requested to run over zero events!");
+    } else {
+      m_nEvents = nEvents;
+    }
+    if (nEventsInFiles < m_nEvents) {
+      m_nEvents = nEventsInFiles;
+    }
+
+    // std::cout << "podio::ROOTDataSource: Running over " << m_nEvents << " events."
+    //           << std::endl;
+
+    // Get collections stored in the files
+    std::vector<std::string> collNames = frame.getAvailableCollections();
+    // std::cout << "podio::ROOTDataSource: Found following collections:\n";
+    for (auto& collName: collNames) {
+      const podio::CollectionBase* coll = frame.get(collName);
+      if (coll->isValid()) {
+        m_columnNames.emplace_back(collName);
+        m_columnTypes.emplace_back(coll->getValueTypeName());
+        // std::cout << "                - " << collName << "\n";
+      }
+    }
+  }
+
+
+  /**
+   * \brief Inform the podio::ROOTDataSource of the desired level of parallelism.
+   */
+  void
+  ROOTDataSource::SetNSlots(unsigned int nSlots) {
+    // std::cout << "podio::ROOTDataSource: Setting num. of slots to: " << nSlots
+    //           << std::endl;
+    m_nSlots = nSlots;
+
+    if (m_nSlots > m_nEvents) {
+      throw std::runtime_error("podio::ROOTDataSource: Number of events too small!");
+    }
+
+    int eventsPerSlot = m_nEvents / m_nSlots;
+    for (size_t i = 0; i < (m_nSlots - 1); ++i) {
+      m_rangesAll.emplace_back(eventsPerSlot * i, eventsPerSlot * (i + 1));
+    }
+    m_rangesAll.emplace_back(eventsPerSlot * (m_nSlots - 1), m_nEvents);
+    m_rangesAvailable = m_rangesAll;
+
+    // Initialize set of addresses needed
+    m_Collections.resize(
+        m_columnNames.size(),
+        std::vector<const podio::CollectionBase*>(m_nSlots, nullptr));
+
+    // Initialize podio readers
+    for (size_t i = 0; i < m_nSlots; ++i) {
+      m_podioReaders.emplace_back(std::make_unique<podio::ROOTReader>());
+    }
+
+    for (size_t i = 0; i < m_nSlots; ++i) {
+      m_podioReaders[i]->openFiles(m_filePathList);
+    }
+
+    for (size_t i = 0; i < m_nSlots; ++i) {
+      m_frames.emplace_back(
+          std::make_unique<podio::Frame>(
+              podio::Frame(m_podioReaders[i]->readEntry("events", 0))));
+    }
```

Comment:
Aren't we reading the same event `nSlots` times this way? As far as I can see we read the correct entry in any case later in `SetEntry`. Can we not just simply put an empty Frame here? Or even just populate the vector with a bunch of `nullptr`?